unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Raycast/Handler/RaycastHandler.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using DCL;
+using DCL.Components;
+using DCL.Controllers;
+using DCL.ECS7;
+using DCL.ECSComponents;
+using DCL.ECSRuntime;
+using DCL.Helpers;
+using DCL.Models;
+using UnityEngine;
+using RaycastHit = DCL.ECSComponents.RaycastHit;
+
+namespace DCLPlugins.ECSComponents.Raycast
+{
+    public class RaycastComponentHandler : IECSComponentHandler<PBRaycast>
+    {
+        private IECSComponentWriter componentWriter;
+        private DataStore_ECS7 dataStore;
+        private LayerMask layerMaskTarget;
+        
+        public RaycastComponentHandler(IECSComponentWriter componentWriter, DataStore_ECS7 dataStore)
+        {
+            this.dataStore = dataStore;
+            this.componentWriter = componentWriter;
+            this.layerMaskTarget = 1 << LayerMask.NameToLayer("OnPointerEvent");
+        }
+        
+        public void OnComponentCreated(IParcelScene scene, IDCLEntity entity)
+        {
+        }
+
+        public void OnComponentRemoved(IParcelScene scene, IDCLEntity entity)
+        {
+        }
+
+        public void OnComponentModelUpdated(IParcelScene scene, IDCLEntity entity, PBRaycast model)
+        {
+            UnityEngine.Vector3 origin = new UnityEngine.Vector3(model.Origin.X, model.Origin.Y, model.Origin.Z);
+            UnityEngine.Ray ray = new UnityEngine.Ray();
+            ray.origin =  origin + Utils.GridToWorldPosition(scene.sceneData.basePosition.x, scene.sceneData.basePosition.y);
+            ray.direction = new UnityEngine.Vector3(model.Direction.X, model.Direction.Y, model.Direction.Z);
+            
+            PBRaycastResult result = new PBRaycastResult();
+            result.Direction = model.Direction.Clone();
+            result.Origin = model.Origin.Clone();
+            result.Timestamp = model.Timestamp;
+            
+            UnityEngine.RaycastHit[] hits = Array.Empty<UnityEngine.RaycastHit>();
+
+            if (model.QueryType == QueryType.HitFirst)
+            {
+                bool hasHit = UnityEngine.Physics.Raycast(ray, out UnityEngine.RaycastHit hit, model.MaxDistance, ~layerMaskTarget);
+                if (hasHit)
+                {
+                    hits = new UnityEngine.RaycastHit[1];
+                    hits[0] = hit;
+                }
+            }
+            else if (model.QueryType == QueryType.QueryAll)
+            {
+                hits = UnityEngine.Physics.RaycastAll(ray, model.MaxDistance, ~layerMaskTarget);
+            }
+            else
+            {
+                return;
+            }
+
+            foreach (var hitInfo in hits)
+            {
+                IDCLEntity collisionEntity = null;
+                foreach (var entityI in scene.entities)
```

Comment:
```suggestion
                foreach (var entityI in scene.entities.Values)
```
you only need the values of the dictionary here. using `.Values` avoid a lot of overhead while iterating